nwb_conversion_tools/utils/basesinwbephyswriter.py

```diff
@@ -0,0 +1,87 @@
+import distutils.version
+from abc import ABC, abstractmethod
+
+import numpy as np
+import pynwb
+
+from .basenwbephyswriter import BaseNwbEphysWriter
+from .common_writer_tools import default_return
+
+
+class BaseSINwbEphysWriter(BaseNwbEphysWriter, ABC):
+    def __init__(self, object_to_write, stub, stub_channels):
+        self.recording, self.sorting, self.waveforms, self.event = None, None, None, None
+        BaseNwbEphysWriter.__init__(self, object_to_write, stub=stub, stub_channels=stub_channels)
+
+    @abstractmethod
+    def _make_recording_stub(self):
+        pass
+
+    @abstractmethod
+    def _make_sorting_stub(self):
+        pass
+
+    # @default_return(None)
+    def _get_sampling_frequency(self):
+        return self.recording.get_sampling_frequency()
+
+    # @default_return([])
+    def _get_channel_ids(self):
+        return self.recording.get_channel_ids()
+
+    # @default_return(None)
+    def _get_unit_sampling_frequency(self):
+        return self.sorting.get_sampling_frequency()
+
+    # @default_return([])
+    def _get_unit_ids(self):
+        return self.sorting.get_unit_ids()
+
+    # @default_return([])
+    def _check_valid_property(self, prop_values):
+        if not isinstance(prop_values[0], tuple(self.dt_column_defaults)):
+            return
+        if isinstance(prop_values[0], np.ndarray):
+            shapes = [value.shape[1:] if len(value.shape) > 1 else 1 for value in prop_values]
+            if np.all([shape == shapes[0] for shape in shapes]):
+                return prop_values
+        else:
+            return prop_values
+
+    def add_recording(self, segment_index=0):
+        assert (
+            distutils.version.LooseVersion(pynwb.__version__) >= "1.3.3"
+        ), "'write_recording' not supported for version < 1.3.3. Run pip install --upgrade pynwb"
+
+        self.add_devices()
+        self.add_electrode_groups()
+        self.add_electrodes()
+        if self._conversion_ops["write_electrical_series"]:
+            self.add_electrical_series(segment_index=segment_index)
+
+    def add_sorting(self):
+        self.add_units()
+
+    def add_waveforms(self):
+        if self.waveforms is not None:
+            self.add_units_waveforms()
+
+    def add_to_nwb(self, nwbfile: pynwb.NWBFile, metadata=None, **kwargs):
```

Comment:
I'd rather have the `nwbfile` and `metadata` passed as arguments to the init. Note that the `nwbfile` can be `None` by default in case one only needs the functions to get proeprties etc. With this implementation, it would be impossible to call separately a `add_***` function because the `metadata` and `nwbfile` would not be instantiated.

Reply:
in the base class [here](https://github.com/catalystneuro/nwb-conversion-tools/blob/584faa8ee7a33187345ae2a85764648aa2b444e0/nwb_conversion_tools/utils/basenwbephyswriter.py#L31) I do instantiate it, but dont require them as args. 
I've moved this to another PR to this branch

Reply:
@alejoe91 please check #270